include/boost/gil/image_processing/threshold.hpp

```diff
@@ -0,0 +1,156 @@
+//
+// Copyright 2019 Miral Shah <miralshah2211@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef BOOST_GIL_IMAGE_PROCESSING_THRESHOLD_HPP
+#define BOOST_GIL_IMAGE_PROCESSING_THRESHOLD_HPP
+
+#include <functional>
+#include <limits>
+#include <boost/gil/image.hpp>
+
+namespace boost { namespace gil {
+
+
+namespace detail {
+
+template
+<
+    typename SourceChannelT,
+    typename ResultChannelT,
+    typename SrcView,
+    typename DstView
+>
+void threshold_simple_impl
+(
+    SrcView const& src_view,
+    DstView const& dst_view,
+    std::function<ResultChannelT(SourceChannelT)> const& logic
+)
+{
+    //template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+    gil_function_requires<ColorSpacesCompatibleConcept
+    <
+        typename color_space_type<SrcView>::type,
+        typename color_space_type<DstView>::type>
+    >();
+
+    //iterate over the image chaecking each pixel value for the threshold
+    for (size_t y = 0; y < src_view.height(); y++)
+    {
+        typename SrcView::x_iterator src_it = src_view.row_begin(y);
+        typename DstView::x_iterator dst_it = dst_view.row_begin(y);
+
+        for (size_t x = 0; x < src_view.width(); x++)
```

Comment:
The `size_t` should be replaced with `std::ptrdiff_t`, that type returned by the `width()`